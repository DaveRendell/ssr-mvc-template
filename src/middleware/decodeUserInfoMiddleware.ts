import {NextFunction, Request, Response} from "express"
import jwt from "jsonwebtoken"
import Config from "../config"

export const cookieName = "user-token"

function isUserInfo(
  decodedToken: string | object
): decodedToken is Express.User {
  return (decodedToken as Express.User).id !== undefined
    && (decodedToken as Express.User).displayName !== undefined
}

export default class DecodeUserInfoMiddleware {
  public config: Config

  constructor(config: Config) {
    this.config = config

    this.decode = this.decode.bind(this)
  }

  public decode(request: Request, response: Response, next: NextFunction) {
    const token = request.cookies[cookieName]

    if (!token) {
      return next()
    }

    let decodedToken
    try {
      decodedToken = jwt.verify(
        token,
        this.config.authentication.jwtSecret
      )
    } catch (error) {
      return next()
    }

    if (isUserInfo(decodedToken)) {
      const {id, displayName} = decodedToken
      request.user = {id, displayName}
    }

    next()
  }
}
