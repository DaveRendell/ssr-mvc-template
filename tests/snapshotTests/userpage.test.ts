import request from "supertest"
import {cookieName} from '../../src/middleware/decodeUserInfoMiddleware'
import {startApp} from "./snapshotTestUtilities"
import {httpStatus, getJwtToken} from "../testUtilities"

describe("User page", () => {
  let app: Express.Application
  const validJwtToken = getJwtToken(1, "test-name")
  const invalidJwtToken = getJwtToken(-1, "non-existant-user")

  beforeAll(async () => {
    app = await startApp()
  })
  
  describe("index", () => {
    it("returns unauthorised if user is not logged in", async () => {
      const response = await request(app).get("/user/")

      expect(response.status).toBe(httpStatus.unauthorized)
      expect(response.text).toMatchSnapshot()
    })

    it("returns unauthorised if user token is not valid", async () => {
      const response = await request(app)
        .get("/user/")
        .set("Cookie", [`${cookieName}=${invalidJwtToken}`])

      expect(response.status).toBe(httpStatus.unauthorized)
      expect(response.text).toMatchSnapshot()
    })

    it("returns user page content if user is logged in", async () => {
      const response = await request(app)
        .get("/user/")
        .set("Cookie", [`${cookieName}=${validJwtToken}`])

      expect(response.status).toBe(httpStatus.ok)
      expect(response.text).toMatchSnapshot()
    })
  })
})
