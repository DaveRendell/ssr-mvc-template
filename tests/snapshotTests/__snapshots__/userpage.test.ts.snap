// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User page index returns unauthorised if user is not logged in 1`] = `"Unauthorized"`;

exports[`User page index returns unauthorised if user token is not valid 1`] = `"Unauthorized"`;

exports[`User page index returns user page content if user is logged in 1`] = `
<html>
  <head>
    <link rel="stylesheet"
          type="text/css"
          href="styles/app.css"
    >
  </head>
  <body>
    <div class="page-header">
      <strong>
        App Header
      </strong>
    </div>
    <div class="page-content">
      <div>
        <h1>
          User Page
        </h1>
        <p>
          Display name: test-name
        </p>
      </div>
    </div>
  </body>
</html>
`;
